network/test/block_exchange_test.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/module/util"
+	"github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/network/p2p"
+)
+
+type BlockExchangeTestSuite struct {
+	suite.Suite
+
+	cancel         context.CancelFunc
+	networks       []*p2p.Network
+	blockExchanges []network.BlockExchange
+	blockCids      []cid.Cid
+	numNetworks    int
+}
+
+func (suite *BlockExchangeTestSuite) SetupTest() {
+	suite.numNetworks = 3
+
+	logger := zerolog.New(os.Stdout)
+
+	_, _, networks, _, cancel := GenerateIDsMiddlewaresNetworks(
+		suite.T(), suite.numNetworks, logger, 100, nil, false, nil, []dht.Option{p2p.AsServer(true)},
+	)
+	suite.networks = networks
+	suite.cancel = cancel
+
+	var blockstores []blockstore.Blockstore
+
+	blockExchangeChannel := network.Channel("block-exchange")
+
+	for i, net := range networks {
+		bstore := makeBlockstore(suite.T(), fmt.Sprintf("bs%v", i))
+		block := blocks.NewBlock([]byte(fmt.Sprintf("foo%v", i)))
+		suite.blockCids = append(suite.blockCids, block.Cid())
+		require.NoError(suite.T(), bstore.Put(block))
+		blockstores = append(blockstores, bstore)
+		bex, err := net.RegisterBlockExchange(blockExchangeChannel, blockstores[i])
+		require.NoError(suite.T(), err)
+		suite.blockExchanges = append(suite.blockExchanges, bex)
+	}
+}
+
+func (suite *BlockExchangeTestSuite) TearDownTest() {
+	suite.cancel()
+	netDoneChans := make([]<-chan struct{}, len(suite.networks))
+	for i, net := range suite.networks {
+		netDoneChans[i] = net.Done()
+	}
+	<-util.AllClosed(netDoneChans...)
+}
+
+func (suite *BlockExchangeTestSuite) TestGetBlocks() {
+	for i, bex := range suite.blockExchanges {
+		// check that we can get all other blocks
+		var blocksToGet []cid.Cid
+		for j, blockCid := range suite.blockCids {
+			if j != i {
+				blocksToGet = append(blocksToGet, blockCid)
+			}
+		}
+
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+
+		blocksReceived := make(map[cid.Cid]struct{})
+		done, err := bex.GetBlocks(blocksToGet...).ForEach(func(b blocks.Block) {
+			blocksReceived[b.Cid()] = struct{}{}
+		}).Send(ctx)
+		require.NoError(suite.T(), err)
+
+		<-done
+		cancel()
+
+		for _, blockCid := range blocksToGet {
+			_, blockReceived := blocksReceived[blockCid]
+			assert.True(suite.T(), blockReceived, "block %v not received by node %v", blockCid, i)
+		}
+	}
+}
+
+func (suite *BlockExchangeTestSuite) TestGetBlocksWithSession() {
+	for i, bex := range suite.blockExchanges {
+		// check that we can get all other blocks in a single session
+		blocksToGet := make(map[cid.Cid]struct{})
+		for j, blockCid := range suite.blockCids {
+			if j != i {
+				blocksToGet[blockCid] = struct{}{}
+			}
+		}
+
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+
+		var doneChans []<-chan struct{}
+		session := bex.GetSession(ctx)
```

Comment:
what does a session signify here? Is this is a BlockSwap concept?

Reply:
Yes, it is a bitswap concept: https://github.com/ipfs/go-bitswap#get-related-blocks-faster-with-sessions